tea.py

```diff
@@ -0,0 +1,674 @@
+                    (1 - pe_porosity)
+                    * pe_composite_density
+                    / (
+                        (1 - pe_porosity) * pe_composite_density
+                        + pe_porosity * elyte_density
+                    )
+                )
+                pe_elyte_wt_fraction = 1 - pe_composite_wt_fraction
+        try:
+            sep_composite_wt_fraction = input_parameter[
+                "Separator composite weight fraction"
+            ]
+            sep_elyte_wt_fraction = 1 - sep_composite_wt_fraction
+        except KeyError:
+            try:
+                sep_elyte_wt_fraction = input_parameter[
+                    "Separator electrolyte weight fraction"
+                ]
+                sep_composite_wt_fraction = 1 - sep_elyte_wt_fraction
+            except KeyError:
+                sep_composite_wt_fraction = (
+                    (1 - sep_porosity)
+                    * sep_density
+                    / ((1 - sep_porosity) * sep_density + sep_porosity * elyte_density)
+                )
+                sep_elyte_wt_fraction = 1 - sep_composite_wt_fraction
+
+        # electrode and separator with electrolyte density
+        try:
+            ne_w_elyte_density = input_parameter[
+                "Negative electrode with electrolyte density [kg.m-3]"
+            ]
+        except KeyError:
+            ne_w_elyte_density = (
+                ne_composite_density * ne_composite_wt_fraction
+                + elyte_density * ne_elyte_wt_fraction
+            )
+        try:
+            pe_w_elyte_density = input_parameter[
+                "Positive electrode with electrolyte density [kg.m-3]"
+            ]
+        except KeyError:
+            pe_w_elyte_density = (
+                pe_composite_density * pe_composite_wt_fraction
+                + elyte_density * pe_elyte_wt_fraction
+            )
+        try:
+            sep_w_density = input_parameter[
+                "Separator with electrolyte density [kg.m-3]"
+            ]
+        except KeyError:
+            sep_w_elyte_density = (
+                sep_composite_density * sep_composite_wt_fraction
+                + elyte_density * sep_elyte_wt_fraction
+            )
+
+        # gravimetric stack energy density wh.kg-1
+        try:
+            stack_density = input_parameter[
+                "Stack density [kg.m-3]"
+            ]  # better 'stack mixed density'?
+        except KeyError:
+            stack_density = (
+                ne_cc_thickness / 2 * ne_cc_density
+                + ne_thickness * ne_w_elyte_density
+                + sep_thickness * sep_w_elyte_density
+                + pe_thickness * pe_w_elyte_density
+                + pe_cc_thickness / 2 * pe_cc_thickness
+            ) / stack_thickness  # [kg.m-3]
+
+        gravimetric_stack_energy_density = (
+            volumetric_stack_energy_density / stack_density * 1000
+        )  # [Wh.kg-1]
+
+        if print_values == True:
```

Comment:
I think the class should have separate methods for calculations that store the results in a dict that is an attribute of the class. E.g. in the `__init__` define `self.results = {}` and update this as you go. Then you can have separate methods for printing and plotting

Reply:
Then printing is a one-liner that just prints key, value pairs in the dict